### PR TITLE
BTC-68: fix cpuminer, add pop_witness_commitment

### PR DESCRIPTION
**Pull request by ghost**
What I've done:
1. Added an increase in cbtx_size after adding data
2. Increased the counter by one (there was a hardcode 1 or 2 when segwit, but you need one higher, because we are adding an output)
3. Fixed hash calculation
4. Fixed reading of keystones (reverse order)

### CODE REVIEW

File: cpu-miner.c [hunk 1]
```diff
@@ -343,23 +343,42 @@ static bool work_decode(const json_t *val, struct work *work)
 	return false;
 }
 
+static void tool_invert_buffer(void* buf, size_t bufSize)
```
Comment (Warchant) [inline cpu-miner.c]: use memrev instead of this

File: cpu-miner.c [hunk 2]
```diff
@@ -343,23 +343,42 @@ static bool work_decode(const json_t *val, struct work *work)
...
 	bool submit_coinbase = false;
 	bool segwit = false;
-	json_t *tmp, *txa;
+	json_t *tmp, *txa, *keystones;
 	bool rc = false;
 
+	// VeriBlock:
+	const int keystones_total = 2;
+	const int unauthcontext_size = 4 /* sizeof(int) */ + keystones_total * 32;
+        unsigned char unauthcontext[unauthcontext_size];
```
Comment (Warchant) [inline cpu-miner.c]:
```suggestion
      unsigned char unauthcontext[unauthcontext_size];
```

File: cpu-miner.c [hunk 3]
```diff
@@ -440,7 +465,8 @@ static bool gbt_work_decode(const json_t *val, struct work *work)
 	if (tmp) {
 		const char *cbtx_hex = json_string_value(json_object_get(tmp, "data"));
 		cbtx_size = cbtx_hex ? strlen(cbtx_hex) / 2 : 0;
-		cbtx = malloc(cbtx_size + 100);
+		cbtx = malloc(cbtx_size + 1000);
+        memset(cbtx, 0, cbtx_size + 1000);
```
Comment (Warchant) [inline cpu-miner.c]: align with other code

File: cpu-miner.c [hunk 4]
```diff
@@ -477,21 +507,45 @@ static bool gbt_work_decode(const json_t *val, struct work *work)
...
-		cbtx[cbtx_size++] = segwit ? 2 : 1; /* out-counter */
+		cbtx[cbtx_size++] = (segwit ? 2 : 1) + 1; /* out-counter */
 		le32enc((uint32_t *)(cbtx+cbtx_size), (uint32_t)cbvalue); /* value */
 		le32enc((uint32_t *)(cbtx+cbtx_size+4), cbvalue >> 32);
 		cbtx_size += 8;
 		cbtx[cbtx_size++] = pk_script_size; /* txout-script length */
 		memcpy(cbtx+cbtx_size, pk_script, pk_script_size);
 		cbtx_size += pk_script_size;
+
+        //=================================================================
```
Comment (Warchant) [inline cpu-miner.c]: add comment: "Adding new output with POP root"

File: cpu-miner.c [hunk 5]
```diff
@@ -573,8 +628,40 @@ static bool gbt_work_decode(const json_t *val, struct work *work)
...
+    for (k=0; k<2; k++)
+    {
+        unsigned char *ks_begin = unauthcontext + 4 + k*keystoneSize;
+        tool_invert_buffer(ks_begin, keystoneSize);
+    }
+
 	/* generate merkle root */
-	merkle_tree = malloc(32 * ((1 + tx_count + 1) & ~1));
+    unsigned long mtreeSize = 32 * ((1 + tx_count + 1) & ~1);
+	merkle_tree = malloc(mtreeSize);
```
Comment (Warchant) [inline cpu-miner.c:650]: do not use tabs, use only spaces

File: cpu-miner.c [hunk 6]
```diff
@@ -573,8 +628,40 @@ static bool gbt_work_decode(const json_t *val, struct work *work)
...
+    memset(merkle_tree, 0, mtreeSize);
```
Comment (Warchant) [inline cpu-miner.c:651]: align with other code

File: cpu-miner.c [hunk 7]
```diff
@@ -611,6 +698,17 @@ static bool gbt_work_decode(const json_t *val, struct work *work)
 			sha256d(merkle_tree[i], merkle_tree[2*i], 64);
 	}
 
+	// VeriBlock: merkle_tree[0] is a btc merkle root
+	// calculate top level merkle root as sha256d(sha256d(unauthcontext) || merkle_root[0])
+    // to do this, copy existing merkle_root[0] into merkle_root[1]
+    // then copy sha256d(unauthcontext) into merkle_root[0]
+    // and do one more round of hashing
```
Comment (Warchant) [inline cpu-miner.c]: align with other code

**Review by Warchant — CHANGES_REQUESTED** [review]

File: cpu-miner.c (hunk 7 above)
Comment (ghost) [reply cpu-miner.c]: done

File: cpu-miner.c (hunk 6 above)
Comment (ghost) [reply cpu-miner.c:651]: done

File: cpu-miner.c (hunk 5 above)
Comment (ghost) [reply cpu-miner.c:650]: all project uses tabs
I delete my spaces and inserted tabls as same as project

File: cpu-miner.c (hunk 4 above)
Comment (ghost) [reply cpu-miner.c]: done

**Review by Warchant — APPROVED** [review]

File: cpu-miner.c (hunk 3 above)
Comment (ghost) [reply cpu-miner.c]: done

File: cpu-miner.c (hunk 2 above)
Comment (ghost) [reply cpu-miner.c]: done

File: cpu-miner.c (hunk 1 above)
Comment (ghost) [reply cpu-miner.c]: done

**Pull request merged**